### PR TITLE
Fix certificate username conflict

### LINKED ISSUE #16490: username and certificate conflict throws an exception and returns 500 http status code
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

authentication

### Describe the bug

If users try to get a token with an incorrect certificate, then:
- an AuthenticationFlowException is thrown:
`ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: org.keycloak.authentication.AuthenticationFlowException at org.keycloak.authentication.AuthenticationProcessor.setAutheticatedUser(AuthenticationProcessor.java:273)`

- HTTP status code 500 is returned:
`{
    "error": "unknown_error"
}`

### Version

9.0.3/16.1.1/19.0.3

### Expected behavior

If users try to get a token with an incorrect certificate, then:

- no Exception is thrown
- HTTP status code 401 is returned:
`{
    "error": "USER_CONFLICT",
    "error_description": "authentication failed, check your credentials"
}`

### Actual behavior

If users try to get a token with an incorrect certificate, then:
- an AuthenticationFlowException is thrown:
`ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: org.keycloak.authentication.AuthenticationFlowException at org.keycloak.authentication.AuthenticationProcessor.setAutheticatedUser(AuthenticationProcessor.java:273)`

- HTTP status code 500 is returned:
`{
    "error": "unknown_error"
}`

### How to Reproduce?

> curl -L -X POST 'https://localhost/auth/realms/the-Realm/protocol/openid-connect/token' 
-H 'Content-Type: application/x-www-form-urlencoded' 
--data-urlencode 'grant_type=the-password' 
--data-urlencode 'client_id=the-client' 
--data-urlencode 'client_secret=the-client-secret' 
--data-urlencode 'username=wrong-username' 
--cert-type P12 --cert the-right-certificate.p12:the-password

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Bukarim**
Closes #16490

### CODE REVIEW

**Review by mposolda — CHANGES_REQUESTED** [review]
@Bukarim I think that KeycloakErrorHandler is not proper place where to handle this case. It should be handled at the place where this error actually happened. Also it is needed to have a test for this.

**Comment by Bukarim** [comment]
@mposolda thx for your feedback, i will change that :)

**Comment by Bukarim** [comment]
@mposolda, I've moved the handling to the TokenEndpoint, I hope it's the right place.

In our company we use certificate-based authentication, we tested the fix and it conforms the expectations.
The subject "testing" is ambiguous/complicated: a certificate that is not outdated is required!! hence my question: is it possible to mock the test? Which package does this test belong to?

**Comment by mposolda** [comment]
@Bukarim Thanks for the changes.

We try to avoid mocks. The "real test" can be nice. There is `X509DirectGrantTest`, which tests X.509 authentication for direct grant login (OAuth2 resource owner password credentials) and hence looks like good place to test this. I think you can relatively easily add new test method to this class. It seems to me that you don't need to generate new certificate or something like that (which is a bit pain, so good that you don't need this) as you can just use `username` parameter for some different user than the user present in the certificate and that's sufficient to reproduce the issue if I understand correctly?

**Comment by Bukarim** [comment]
Hi @mposolda,
your tip was very helpful, thank you!
yes, you understood the situation very well.
I wrote a test for the use case, but surprisingly it failed: a token is always generated and the username doesn't matter, only the clientSecret has to be right.
I wrote a temporary method (org.keycloak.testsuite.util.OAuthClient#doAccessTokenRequestTemp) hoping to get the expected result, but it doesn't help.
is my test set wrong?
Any idea why the method (oauth.doGrantAccessTokenRequest) always returns a token even if the username is wrong?

**Review by vilmosnagy — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java [hunk 1]
```diff
@@ -635,7 +640,15 @@ public Response resourceOwnerPasswordCredentialsGrant() {
                 .setSession(session)
                 .setUriInfo(session.getContext().getUri())
                 .setRequest(request);
-        Response challenge = processor.authenticateOnly();
+        Response challenge;
+        try{
+            challenge = processor.authenticateOnly();
+        }catch (AuthenticationFlowException exception){
```
[the remark this answers is not in the document]
Comment (Bukarim) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:646]: i adjusted the try catch block 👍

**Comment by Bukarim** [comment]
Hi @mposolda @vilmosnagy @stianst @rmartinc, could someone answer reply to my [comment](https://github.com/keycloak/keycloak/pull/16499#issuecomment-1443898247)
i need help to fix the test: (org.keycloak.testsuite.x509.X509DirectGrantTest#loginFailedOnCertificateCNIsNotTheSameAsClientUsername).

the method (org.keycloak.testsuite.util.OAuthClient#doGrantAccessTokenRequest(java.lang.String, java.lang.String, java.lang.String)) always returns a token, even if the username doesn't match the CN in the used certificate.

**Comment by jonkoops** [comment]
@Bukarim could you rebase your work on `main`? There are a ton of unrelated commits in this PR at the moment which makes it hard to understand what changed.

**Comment by rmartinc** [comment]
@Bukarim I think that the test works because the direct grant flow configuration for the test just contains the "X509/Validate Username" step (and not "Username Validation"). In my tests I just got your exception if the flow contains both steps (first "Username Validation" and then "X509/Validate Username", both required). Note that "Username Validation" is not needed for the X509 login, the user is obtained from the X509 certificate following the cx509 configuration. Is your direct grant flow containing the "Username Validation" step too?

**Comment by Bukarim** [comment]
we've adjusted our flow and it's working as expected 🥇

**Pull request closed without merging**